batcheval: add WriteOptions struct to batch request header

msbutler opened the pull request:
To prevent data looping in Logical Data Replication, the LDR writer will bind
an OriginID to each batch request it sends. This patch plumbs the OriginID in
the header to the MVCCWriteOptions, which will be used to bind the OriginID to
each commited KV's MVCCValueHeader.

This patch adds a WriteOptions struct to the request header because future
options may get added to this container, like the OriginTimestamp.

Informs https://github.com/cockroachdb/cockroach/issues/126253

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126285)
<!-- Reviewable:end -->

msbutler committed:
batcheval: add WriteOptions struct to batch request header

To prevent data looping in Logical Data Replication, the LDR writer will bind
an OriginID to each batch request it sends. This patch plumbs the OriginID in
the header to the MVCCWriteOptions, which will be used to bind the OriginID to
each commited KV's MVCCValueHeader.

This patch adds a WriteOptions struct to the request header because future
options may get added to this container, like the OriginTimestamp.

Informs #126253

Release note: none

msbutler commented:
I was thinking of adding unit tests once I had hooked up the sql plumbing, but if I should add tests here, I'd appreciate a pointer to a similar test.

miraradeva reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126285#-:-O0U1rgq8uaFzy3vEk3i:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt and @msbutler)

___
*[`pkg/kv/kvpb/api.go` line 2537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126285#-O0TzHRX6q4Zb31Wj1SH:-O0TzHRYD8h15sOiyhZk:b-pfndve) ([raw file](https://github.com/cockroachdb/cockroach/blob/59ba770e75a2f85cf11efc2fafe64ad0bd8cbbbc/pkg/kv/kvpb/api.go#L2537)):*
> ```Go
> 
> func (writeOptions *WriteOptions) GetOriginID() uint32 {
> 	if writeOptions == nil {
> ```

This should be `if writeOptions.OriginID == nil`, right?


<!-- Sent from Reviewable.io -->

miraradeva commented:
> I was thinking of adding unit tests once I had hooked up the sql plumbing, but if I should add tests here, I'd appreciate a pointer to a similar test.

The only part I can think of testing is the nullable nature of the new `WriteOptions` but I don't see any similar tests for the other nullable fields of `Header`.

msbutler reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126285)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dt and @miraradeva)

___
*[`pkg/kv/kvpb/api.go` line 2537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126285#-O0TzHRX6q4Zb31Wj1SH:-O0U3cDIDzrIC42kjzCP:b-ps0169) ([raw file](https://github.com/cockroachdb/cockroach/blob/59ba770e75a2f85cf11efc2fafe64ad0bd8cbbbc/pkg/kv/kvpb/api.go#L2537)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

This should be `if writeOptions.OriginID == nil`, right?
</blockquote></details>

nah, this was intended. In the call site:
```
h.WriteOptions.GetOriginID()
```

I want to avoid a null pointer exception on WriteOptions when I grab OriginID.


<!-- Sent from Reviewable.io -->

msbutler commented:
Ah, I'll be able to test this in a follow up PR in a test similar to `TestReplicaRangefeed`

msbutler commented:
TFTRs!

bors r=miraradeva,dt
